src/lib/blocks.js

```diff
@@ -110,6 +110,75 @@ export default function (vm) {
+        const size = ScratchBlocks.ScratchMsgs.translate('SENSING_OF_SIZE', 'size');
+        const volume = ScratchBlocks.ScratchMsgs.translate('SENSING_OF_VOLUME', 'volume');
+        let output = [
+            [xPosition, 'x position'],
+            [yPosition, 'y position'],
+            [direction, 'direction'],
+            [costumeNumber, 'costume #'],
+            [costumeName, 'costume name'],
+            [size, 'size'],
+            [volume, 'volume']
+        ];
+
+        const sourceBlock = thisValue.sourceBlock_; // This is the <shadow>.
+        const ofBlock = sourceBlock && sourceBlock.parentBlock_; // This is the "of" block.
+        if (ofBlock) {
+            let block;
+            let blocks;
+            if (vm.editingTarget) {
+                blocks = vm.editingTarget.blocks;
+                block = blocks.getBlock(ofBlock.id);
+            }
+            if (!block) {
+                // The block may be in the flyout.
+                blocks = vm.runtime.flyoutBlocks;
+                block = blocks.getBlock(ofBlock.id);
+            }
+            if (block) {
+                // Get the name of the sprite which is selected. This is based on the OBJECT dropdown's selected value,
+                // which may be different from the value returned by a block placed within that input, if present.
+                // TODO: Variables should probably not be fetched at all if the OBJECT dropdown contains a block.
+                const objectInput = blocks.getInputs(block).OBJECT;
+                const objectInputBlock = blocks.getBlock(objectInput.block);
+                const valueField = blocks.getFields(objectInputBlock).OBJECT;
+                const objectName = valueField.value;
+
+                // Only return options for local variables of sprites, since the stage's local variables are really
+                // the project's global variables.
+                if (objectName === '_stage_') {
+                    // If the stage is the selected object (of the dropdown), only return properties relevant to it.
+                    const backdropNumber = ScratchBlocks.ScratchMsgs.translate(
+                        'SENSING_OF_BACKDROPNUMBER', 'backdrop #');
+                    const backdropName = ScratchBlocks.ScratchMsgs.translate(
+                        'SENSING_OF_BACKDROPNAME', 'backdrop name');
+                    output = [
+                        [backdropNumber, 'backdrop #'],
+                        [backdropName, 'backdrop name'],
+                        [volume, 'volume']
+                    ];
+                } else {
+                    const target = vm.runtime.getSpriteTargetByName(objectName);
+                    if (target) {
+                        // Pass true to skip the stage: we only want the sprite's own local variables.
+                        const variableNames = target.getAllVariableNamesInScopeByType('', true);
```

Comment:
This uses a function which wasn't particularly designed for this purpose, but it's perfectly appropriate for the situation. It returns an array of strings which are the names of the local variables of the sprite - exactly what we need to build the menu options.